stitchee/concatenate.py

```diff
@@ -139,6 +140,104 @@ def validate_concat_method_and_dim(concat_method: str, concat_dim: str | None =
+    # update coordinates with newly renamed dimensions
+    new_coords = {
+        new_dim: dataarray.coords[old_dim].rename({old_dim: new_dim})
+        for old_dim, new_dim in dim_tuple
+        if old_dim in dataarray.coords
+    }
+
+    # construct new DataArray with repeated dimensions
+    new_dataarray = xr.DataArray(
+        dataarray.to_numpy(),
+        dims=new_dims,
+        coords=new_coords,
+        name=dataarray.name,
+        attrs=dataarray.attrs,
+    )
+
+    return new_dataarray
+
+
+def concat_manually(dataarrays, concat_dim: str, concat_kwargs: dict):
+    # Build base kwargs
+    base_kwargs = {**DEFAULT_XARRAY_SETTINGS, **concat_kwargs}
+    # data_vars must be "all" when concatenating DataArrays because there is only one data_var, DataArray itself
+    base_kwargs.update({"data_vars": "all"})
+
+    # construct a list of dataarrays with repeated dimensions renamed
+    new_dataarrays = [rename_to_uniq_dimensions(da) for da in dataarrays]
+    # concatenate the manually constructed dataarrays
+    concatenated_dataarray = xr.concat(new_dataarrays, dim=concat_dim, **base_kwargs)
+
+    # reconstruct the old list of dimensions and coordinates with repeated dims
+    old_dims = [dim.split("___", 1)[-1] for dim in concatenated_dataarray.dims]
+    dim_tuple = list(zip(old_dims, concatenated_dataarray.dims))
+    old_coords = {
+        old_dim: concatenated_dataarray.coords[new_dim].rename({new_dim: old_dim})
+        for old_dim, new_dim in dim_tuple
+        if new_dim in concatenated_dataarray.coords
+    }
+
+    # construct DataArray again with original dimension schema
+    dataarray_back = xr.DataArray(
+        concatenated_dataarray.to_numpy(),
+        dims=old_dims,
+        coords=old_coords,
+        name=concatenated_dataarray.name,
+        attrs=concatenated_dataarray.attrs,
+    )
+    # preserve encoding for storage in NetCDF4
+    dataarray_back.encoding = dataarrays[0].encoding
+
+    return dataarray_back
+
+
+def composite_concat(datasets, concat_dim: str, concat_kwargs: dict):
+    # Build base kwargs
+    base_kwargs = {**DEFAULT_XARRAY_SETTINGS, **concat_kwargs}
+
+    # Find all variables with duplicate dimensions using first dataset
+    duplicate_variables = find_variables_with_duplicate_dimensions(datasets[0])
+
+    # Concatenate datarrays manually for each variable with duplicate dimensions
+    if duplicate_variables:
+        cleaned_datasets = [dataset.drop_vars(duplicate_variables) for dataset in datasets]
+        concatenated_dataset = xr.concat(cleaned_datasets, dim=concat_dim, **base_kwargs)
+        for variable in duplicate_variables:
+            dataarray = concat_manually(
+                [dataset[variable] for dataset in datasets], concat_dim, concat_kwargs
+            )
+            concatenated_dataset = concatenated_dataset.assign({variable: dataarray})
+    else:
+        concatenated_dataset = xr.concat(datasets, dim=concat_dim, **base_kwargs)
```

Comment:
Pulling out the `else` condition to reduce nesting logic:

```suggestion
    if not duplicate_variables:
        return xr.concat(datasets, dim=concat_dim, **base_kwargs)
        
    # Concatenate datarrays first without duplicate dimensions
    clean_datasets = [dataset.drop_vars(duplicate_variables) for dataset in datasets]
    concatenated_dataset = xr.concat(clean_datasets, dim=concat_dim, **base_kwargs)
    
    # Process each duplicate variable separately
    for variable in duplicate_variables:
        dataarray = concat_with_duplicate_dims(
            [dataset[variable] for dataset in datasets], concat_dim, concat_kwargs
        )
        concatenated_dataset = concatenated_dataset.assign({variable: dataarray})
```